File: Sources/PhotoAccessibilityStudio/Services/XMPPacketBuilder.swift
import Foundation

enum XMPPacketError: LocalizedError {
    case malformedPacket
    case directFieldMissing
    case nestedFieldPresent

    var errorDescription: String? {
        switch self {
        case .malformedPacket: return "原始 XMP packet 结构无法安全修改"
        case .directFieldMissing: return "XMP 中缺少直接的 ArtworkContentDescription 元素"
        case .nestedFieldPresent: return "XMP 中仍存在错误的 AOContentDescription 嵌套元素"
        }
    }
}

struct XMPPacketBuilder {
    static let namespace = "http://iptc.org/std/Iptc4xmpExt/2008-02-29/"
    static let directElement = "Iptc4xmpExt:ArtworkContentDescription"
    static let nestedElement = "Iptc4xmpExt:AOContentDescription"

    func build(existing: Data?, description: String) throws -> Data {
        var xml = existing.flatMap { String(data: $0, encoding: .utf8) } ?? ""
        if xml.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            xml = minimalPacket
        }
        xml = removingElements(named: "ArtworkContentDescription", from: xml)
        xml = removingElements(named: "AOContentDescription", from: xml)
        guard let openingRange = xml.range(of: #"<rdf:Description\b[^>]*>"#,
                                           options: .regularExpression) else {
            throw XMPPacketError.malformedPacket
        }
        var opening = String(xml[openingRange])
        let namespacePattern = #"\s+xmlns:Iptc4xmpExt\s*=\s*("[^"]*"|'[^']*')"#
        if let regex = try? NSRegularExpression(pattern: namespacePattern) {
            let range = NSRange(opening.startIndex..<opening.endIndex, in: opening)
            opening = regex.stringByReplacingMatches(in: opening, range: range, withTemplate: "")
        }
        let suffix = opening.hasSuffix("/>") ? "/>" : ">"
        opening.removeLast(suffix.count)
        opening += " xmlns:Iptc4xmpExt=\"\(Self.namespace)\"\(suffix)"
        xml.replaceSubrange(openingRange, with: opening)
        let element = "<\(Self.directElement)>\(escape(description))</\(Self.directElement)>"
        if opening.hasSuffix("/>") {
            guard let refreshed = xml.range(of: #"<rdf:Description\b[^>]*/>"#,
                                            options: .regularExpression) else {
                throw XMPPacketError.malformedPacket
            }
            var expanded = String(xml[refreshed])
            expanded.removeLast(2)
            expanded += ">\n\(element)\n</rdf:Description>"
            xml.replaceSubrange(refreshed, with: expanded)
        } else if let closing = xml.range(of: "</rdf:Description>") {
            xml.insert(contentsOf: "\n\(element)\n", at: closing.lowerBound)
        } else {
            throw XMPPacketError.malformedPacket
        }
        let data = Data(xml.utf8)
        try validateRaw(data)
        return data
    }

    func validateRaw(_ data: Data) throws {
        guard let xml = String(data: data, encoding: .utf8),
              xml.contains("xmlns:Iptc4xmpExt=\"\(Self.namespace)\"") ||
              xml.contains("xmlns:Iptc4xmpExt='\(Self.namespace)'") else {
            throw XMPPacketError.directFieldMissing
        }
        let directPattern = #"<Iptc4xmpExt:ArtworkContentDescription(?:\s[^>]*)?>[\s\S]*?</Iptc4xmpExt:ArtworkContentDescription>"#
        guard xml.range(of: directPattern, options: .regularExpression) != nil else {
            throw XMPPacketError.directFieldMissing
        }
        guard !xml.contains("<\(Self.nestedElement)") else {
            throw XMPPacketError.nestedFieldPresent
        }
    }

    private func removingElements(named localName: String, from xml: String) -> String {
        let pattern = #"<(?:[A-Za-z_][\w.-]*:)?"# + localName +
            #"\b[^>]*>[\s\S]*?</(?:[A-Za-z_][\w.-]*:)?"# + localName + #">"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return xml }
        let range = NSRange(xml.startIndex..<xml.endIndex, in: xml)
        return regex.stringByReplacingMatches(in: xml, range: range, withTemplate: "")
    }

    private func escape(_ value: String) -> String {
        value.replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
            .replacingOccurrences(of: "\"", with: "&quot;")
            .replacingOccurrences(of: "'", with: "&apos;")
    }

    private var minimalPacket: String {
        """
        <?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
        <x:xmpmeta xmlns:x="adobe:ns:meta/">
          <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
            <rdf:Description rdf:about=""></rdf:Description>
          </rdf:RDF>
        </x:xmpmeta>
        <?xpacket end="w"?>
        """
    }
}

File: androidApp/app/src/main/java/com/mingkong/photoaccessibility/XmpPacketBuilder.kt
package com.mingkong.photoaccessibility

object XmpPacketBuilder {
    const val namespace = "http://iptc.org/std/Iptc4xmpExt/2008-02-29/"
    private const val direct = "Iptc4xmpExt:ArtworkContentDescription"
    private const val nested = "Iptc4xmpExt:AOContentDescription"
    private const val xpacketBegin = "\uFEFF"

    fun build(existing: String?, description: String): String {
        require(description.isNotBlank()) { "描述不能为空" }
        require(description.none { it == '\u0000' || (it.code < 32 && !it.isWhitespace()) }) {
            "描述包含无法写入 XML 的控制字符"
        }
        var xml = existing.orEmpty()
        if (xml.isBlank()) xml = emptyPacket()
        xml = removeElement(xml, "ArtworkContentDescription")
        xml = removeElement(xml, "AOContentDescription")

        val openingRegex = Regex("<rdf:Description\\b[^>]*>")
        val openingMatch = openingRegex.find(xml)
            ?: error("原始 XMP packet 缺少 rdf:Description，无法安全修改")
        var opening = openingMatch.value.replace(
            Regex("\\s+xmlns:Iptc4xmpExt\\s*=\\s*(\"[^\"]*\"|'[^']*')"),
            ""
        )
        val suffix = if (opening.endsWith("/>")) "/>" else ">"
        opening = opening.dropLast(suffix.length) + " xmlns:Iptc4xmpExt=\"$namespace\"$suffix"
        xml = xml.replaceRange(openingMatch.range, opening)

        val element = "<$direct>${escape(description)}</$direct>"
        val refreshed = openingRegex.find(xml)
            ?: error("无法定位更新后的 rdf:Description")
        xml = if (refreshed.value.endsWith("/>")) {
            val expanded = refreshed.value.dropLast(2) + ">\n$element\n</rdf:Description>"
            xml.replaceRange(refreshed.range, expanded)
        } else {
            xml.substring(0, refreshed.range.last + 1) + "\n$element\n" +
                xml.substring(refreshed.range.last + 1)
        }
        validate(xml, description)
        return xml
    }

    fun validate(raw: String, expected: String) {
        require(raw.contains("xmlns:Iptc4xmpExt=\"$namespace\"") ||
            raw.contains("xmlns:Iptc4xmpExt='$namespace'")) { "XMP 命名空间不正确" }
        require(!raw.contains("<$nested")) { "XMP 中仍存在错误的 AOContentDescription" }
        val matches = Regex("<$direct(?:\\s[^>]*)?>([\\s\\S]*?)</$direct>").findAll(raw).toList()
        require(matches.size == 1) { "XMP 必须恰好包含一个直接描述字段" }
        require(unescape(matches.single().groupValues[1]) == expected) { "XMP 描述回读不一致" }
    }

    fun extract(raw: String): String? {
        val match = Regex("<$direct(?:\\s[^>]*)?>([\\s\\S]*?)</$direct>").find(raw) ?: return null
        return unescape(match.groupValues[1])
    }

    private fun removeElement(xml: String, localName: String): String = xml.replace(
        Regex(
            "<(?:[A-Za-z_][\\w.-]*:)?$localName\\b[^>]*>[\\s\\S]*?</(?:[A-Za-z_][\\w.-]*:)?$localName>"
        ),
        ""
    )

    private fun escape(value: String) = value.replace("&", "&amp;").replace("<", "&lt;")
        .replace(">", "&gt;").replace("\"", "&quot;").replace("'", "&apos;")

    private fun unescape(value: String) = value.replace("&lt;", "<").replace("&gt;", ">")
        .replace("&quot;", "\"").replace("&apos;", "'").replace("&amp;", "&")

    private fun emptyPacket() = """<?xpacket begin="$xpacketBegin" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/"><rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
<rdf:Description rdf:about=""></rdf:Description>
</rdf:RDF></x:xmpmeta><?xpacket end="w"?>"""
}
